pkg/render/dex_config.go

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) 2019 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package render
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+
+	oprv1 "github.com/tigera/operator/api/v1"
+)
+
+type connectorType string
+
+const (
+	connectorTypeOIDC      = "oidc"
+	connectorTypeOpenshift = "openshift"
+	connectorTypeGoogle    = "google"
+
+	// Various annotations to keep the pod up-to-date
+	authenticationAnnotation = "hash.operator.tigera.io/tigera-dex-auth"
+	dexIdpSecretAnnotation   = "hash.operator.tigera.io/tigera-idp-secret"
+	dexSecretAnnotation      = "hash.operator.tigera.io/tigera-dex-secret"
+	dexTLSSecretAnnotation   = "hash.operator.tigera.io/tigera-dex-tls-secret"
+
+	// Constants related to secrets.
+	serviceAccountSecretField = "serviceAccountSecret"
+	ClientSecretSecretField   = "clientSecret"
+	adminEmailSecretField     = "adminEmail"
+	RootCASecretField         = "rootCA"
+)
+
+// DexConfig is a config for DexIdP itself.
+type DexConfig interface {
+	// UsernameClaim returns the part of the JWT that represents a unique username.
+	UsernameClaim() string
+	// GroupsClaim returns the part of the JWT that represents the list of user groups.
+	GroupsClaim() string
+	// The issuer URL of the upstream IdP
+	IssuerURL() string
+	ConnectorType() string
+	DexKeyValidatorConfig
+}
+
+// DexKeyValidatorConfig is a config for (backend) servers that validate JWTs issued by Dex.
+type DexKeyValidatorConfig interface {
+	// RequiredEnv returns env that is used to configure pods with dex options.
+	RequiredEnv(prefix string) []corev1.EnvVar
+	// RequiredVolumes returns volumes that are related to dex.
+	RequiredVolumes() []corev1.Volume
+	// RequiredVolumeMounts returns volume mounts that are related to dex.
+	RequiredVolumeMounts() []corev1.VolumeMount
+	baseConfig
+}
+
+// DexRelyingPartyConfig is a config for relying parties / applications that use Dex as their IdP.
+type DexRelyingPartyConfig interface {
+	// ClientSecret returns the secret for Dex' auth endpoint
+	ClientSecret() []byte
+	// ManagerURI returns the address where the Manager UI can be found. Ex: https://example.org
+	RequestedScopes() []string
+	// UsernameClaim returns the part of the JWT that represents a unique username.
+	UsernameClaim() string
+	// GroupsClaim returns the part of the JWT that represents the list of user groups.
+	GroupsClaim() string
+	baseConfig
+}
+
+type baseConfig interface {
```

Comment:
Could you add a more specific name to this please, what exactly this is a base of. This is a package of many different types, not specific to dex so anything else in this package could use this.

I also think that if you're going to make some sort of base config it would be much more useful to have a base that was something like this:

```
type K8sRenderer interface {
   RequiredAnnotations() map[string]string
   RequiredSecrets(namespace string) []*corev1.Secret
   RequiredEnv(prefix string) []corev1.EnvVar
   RequiredVolumes() []corev1.Volume
   RequiredVolumeMounts() []corev1.VolumeMount
}
```

Maybe think of a better name then K8sRenderer but the concept is that anything can implement this. I like that you're using interfacing and compositions of interfaces and structs, but I think the concept of the interfaces and the structure needs to be considered first, over just trying to minimise duplicating code.

For instance, is DexConfig really a composition DexKeyValidatorConfig? Does anything would set up the dex config also be a KeyValidator, or have you combining them because they have similar functions? I think both the DexConfig and the DexKeyValidator should implement the K8sRenderer interface, and if you _really_ wanted to add a common interface for the ManagerURI() then create a separate dex specific interface, and compose multiple interfaces in the DexConfig and DexKeyValidatorConfig.